File: Least_squares/Esselman_Least_squares.cpp

#include <iostream>
#include <fstream>
#include <string>
#include <cstring> 
#include <vector>
#include <cmath>
#include <cstdlib>
#include <time.h>
#include "/Users/christopheresselman/vscode_header_files/eigen-3.4.0/Eigen/Dense"
#include <sstream>

using namespace std;
using namespace Eigen;

// Codeium auto completion for C++

//Function to read in CSV and put into an Eigen matrix I got from here:
// https://www.youtube.com/watch?v=K9QB1LbemnY

/*
 * Reads data from a CSV file and returns it as an Eigen Matrix.
 *
 * @param fileToOpen The path to the CSV file to be opened.
 *
 * @return A MatrixXd object containing the data from the CSV file.
 *
 * @throws None.
 */
MatrixXd openData(string fileToOpen)
{
	//Vector to map to eigenmatrix
	vector<double> matrixEntries;

	//Store the data
	ifstream matrixDataFile(fileToOpen);

	if (!matrixDataFile.is_open())
	{
		cerr << "File 1 is not open \n" << "Usage: ./Esselman_Least_squares points.csv \n";
		exit(1);
	}

	//store the row
	string matrixRowString;

	//store matrix entry;
	string matrixEntry;

	//Track number of rows
	int matrixRowNumber = 0;

	//Read in all of the lines into a vector
	while(getline(matrixDataFile,matrixRowString))
	{
		stringstream matrixRowStringStream(matrixRowString);

		while (getline(matrixRowStringStream,matrixEntry,','))
		{
			matrixEntries.push_back(stod(matrixEntry));
		}
		matrixRowNumber++;
	}

	// Number of columns is matrixEntris.size()/matrixRowNumber
	return Map<Matrix<double,Dynamic,Dynamic,RowMajor> >(matrixEntries.data(),matrixRowNumber,matrixEntries.size()/matrixRowNumber);
}

int main(int argc, char **argv)
{
    if (argc != 2)
	{
		cout << "Usage: ./Esselman_Least_squares points.csv \n"; 
		return 1;
	}

	//Save the name of the csv as a string
	string csv_file_name = argv[1];

	//Use the function to read in the csv file
    MatrixXd points = openData(csv_file_name);

	//Make column matrix that is the second column of points
	MatrixXd b = points.col(1);

	//Make a 2 column matrix where the first column is all ones and the second column is the first column of points
	MatrixXd A = MatrixXd::Ones(points.rows(),2);
	A.col(1) = points.col(0);

	//Print to the terminal b and A
	cout << "b = \n" << b << "\n";
	cout  << "A = \n" << A << "\n";

	cout << "points = \n"  << points << "\n";

	//Solve the least squares problem
	MatrixXd x = (A.transpose()*A).inverse()*A.transpose()*b;
	cout << "x = \n" << x << "\n";
    return 0;
}

File: Least_squares/Esselman_Least_squares.py
# Import necessary libraries
import csv  # for reading CSV files
import numpy as np  # for numerical operations
import matplotlib.pyplot as plt  # for plotting

# Prompt the user for the CSV file name
filename = input("Enter CSV file name: ")

# Open the CSV file and read its contents
with open(filename, 'r') as csvfile:
    # Create a CSV reader object
    reader = csv.reader(csvfile)
    # Read the contents of the CSV file into a list of lists (data)
    data = list(reader)

# Make a matrix where the first column is all ones and the second column is the first column of data
A = np.column_stack((np.ones(len(data)), np.array(data)[:, 0]))

# Make a one column matrix from the second column of data
b = np.array(data)[:, 1]

# Make the values of A and b floats
A = A.astype(float)
b = b.astype(float)

# Print A and b
print("A = ", A)
print("b = ", b)

# Solve the least squares problem
x = np.linalg.lstsq(A, b, rcond=None)[0]

# Print the solution
print("x = ", x)
    
